source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -117,115 +185,153 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
   return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
 }
 
-Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
-                                                     bool end_stream) {
-  if (filter_state_ == FilterState::DecodeServingFromCache) {
-    // This call was invoked during decoding by decoder_callbacks_->encodeHeaders because a fresh
-    // cached response was found and is being added to the encoding stream -- ignore it.
-    return Http::FilterHeadersStatus::Continue;
-  }
+void CacheFilter::UpstreamRequest::abort() {
+  stream_->reset(); // Calls onReset, resulting in deletion.
+}
 
-  // If lookup_ is null, the request wasn't cacheable, so the response isn't either.
-  if (!lookup_) {
-    return Http::FilterHeadersStatus::Continue;
+CacheFilter::UpstreamRequest::~UpstreamRequest() {
+  if (filter_ != nullptr) {
+    filter_->onUpstreamRequestReset();
   }
-
-  if (lookup_result_ == nullptr) {
-    // Filter chain iteration is paused while a lookup is outstanding, but the filter chain manager
-    // can still generate a local reply. One case where this can happen is when a downstream idle
-    // timeout fires, which may mean that the HttpCache isn't correctly setting deadlines on its
-    // asynchronous operations or is otherwise getting stuck.
-    ENVOY_BUG(Http::Utility::getResponseStatus(headers) !=
-                  Envoy::enumToInt(Http::Code::RequestTimeout),
-              "Request timed out while cache lookup was outstanding.");
-    filter_state_ = FilterState::NotServingFromCache;
-    return Http::FilterHeadersStatus::Continue;
+  if (lookup_) {
+    lookup_->onDestroy();
+    lookup_ = nullptr;
   }
+  if (insert_queue_) {
```

Comment:
Does insert_queue_ guarantee that all callbacks happen on this thread?

Reply:
Kind of, that's why everything has to be dispatcher-posted. :)